SI-8219 Avoid unpositioned ambiguous implicit error

retronym opened the pull request:
And characterization of the curiously case of overloaded ==

Review by @lrytz

retronym committed:
SI-8219 Pending test case for unpositioned implicit error

Extracted with tweezers from ScalaTest and Scalacheck implicits.

    % scalac-hash v2.11.0-M7 test/pending/pos/t8219.scala
    error: type mismatch;
     found   : Any
     required: AnyRef
    Note: Any is not implicitly converted to AnyRef.  You can safely
    pattern match `x: AnyRef` or cast `x.asInstanceOf[AnyRef]` to do so.
    error: type mismatch;
     found   : Any
     required: AnyRef
    Note: Any is not implicitly converted to AnyRef.  You can safely
    pattern match `x: AnyRef` or cast `x.asInstanceOf[AnyRef]` to do so.
    two errors found

retronym committed:
SI-8219 Avoid leaking ambiguous implicit errors from `viewExists`

This method is called during applicability checking, both to
determine if the given arguments are applicable to a candidate,
and to determine which overload to choose.

This was done using the current types Context. The call to
`inferImplicit` disabled immediate reporting of ambiguous implicit
errors, presumably as these should just be treated as a failed
implicit lookup.

But, in some cases, the buffered errors *were* later issued. Worse
still, they were issued at the position of the provided `EmptyTree`,
rather than at the call site being typechecked.

This commit uses silent mode to completely isolate the results of the
implicit search from the caller's Context.

retronym committed:
SI-8219 Pending test to show suspicous overload in == in AnyRef

    % scalac-hash v2.11.0-M7 test/pending/neg/t8219-any-any-ref-equals.scala
    test/pending/neg/t8219-any-any-ref-equals.scala:5: error: overloaded method value == with alternatives:
      (x$1: AnyRef)Boolean <and>
      (x$1: Any)Boolean
     does not take type parameters
      "".==[Int]
         ^
    one error found

Usually for Java-originated methods, we are allow Object and Any
to unify pre-erasure in method signatures. This is handled in
`matchesParams`, and is predicated on whether the method type is
an instance of `JavaMethodType`

For instance, we are allowed to:

    scala> class C { override def equals(a: Any) = false }
    defined class C

On account of:

    scala> typeOf[Object].decl(nme.equals_).info.getClass
    res7: Class[_ <: $r.intp.global.Type] = class scala.reflect.internal.Types$JavaMethodType

But:

    scala> typeOf[Object].decl(nme.EQ).defString
    res8: String = final def ==(x$1: AnyRef): Boolean

    scala> typeOf[Object].decl(nme.EQ).info.getClass
    res9: Class[_ <: $r.intp.global.Type] = class scala.reflect.internal.Types$MethodType

More special casing is probably needed.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -191,7 +191,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
       && !to.isError
       && context.implicitsEnabled
-      && (inferView(EmptyTree, from, to, reportAmbiguous = false) != EmptyTree)
+      && silent(_.inferView(EmptyTree, from, to, reportAmbiguous = false)).fold(false)(_ != EmptyTree) // SI-8219 silent mode avoids leaking ambiguous errors
```
Isn't it a bug in `inferImplicit`, in that it doesn't respect `reportAmbiguous = false`?

retronym replied:
It doesn't report them immediately.

```
% scalac-hash v2.11.0-M7 -Ylog:typer -Ydebug test/files/pos/t8219.scala 2>&1 | grep update
[log typer] update buffer: Vector([Type error at:NoPosition] type mismatch;
```

``` scala
    if (result.isFailure && saveAmbiguousDivergent && implicitSearchContext.hasErrors) {
      context.updateBuffer(implicitSearchContext.reportBuffer.errors.collect {
        case dte: DivergentImplicitTypeError => dte
        case ate: AmbiguousImplicitTypeError => ate
      })
      debuglog("update buffer: " + implicitSearchContext.reportBuffer.errors)
    }
```

I'm going to try a slightly different approach with `saveErrors = false`.

retronym replied:
pushed

retronym committed:
SI-8219 A less roundabout fix for leaking ambigous errors

I took a second look at this after some scrutiny of the preceding
commit, and found that we really want to turn `saveErrors` off
in `viewExists`. That's a lighter weight than silent mode.

That parameter is passed as `saveAmbigousDivergent` to
`inferImplicit`. If that is enabled, any buffered ambiguous or
divergent implicit errors are appended to the outer typer context.
Which is exactly what we *don't* want to happen here!

lrytz commented:
LGTM!

retronym commented:
Hmm, looks like I'm walking a fine line here:

```
% cat test/files/pos/t8219b.scala
trait Arr[T]
object Arr {
  def apply[T](xs: T*)(implicit evidence$2: scala.reflect.ClassTag[T]): Arr[T] = ???
  def apply(x: Boolean,xs: Boolean*): Arr[Boolean]= ???
  def apply(x: Byte,xs: Byte*): Arr[Byte]= ???
  def apply(x: Short,xs: Short*): Arr[Short]= ???
  def apply(x: Char,xs: Char*): Arr[Char]= ???
  def apply(x: Int,xs: Int*): Arr[Int]= ???
  def apply(x: Long,xs: Long*): Arr[Long]= ???
  def apply(x: Float,xs: Float*): Arr[Float]= ???
  def apply(x: Double,xs: Double*): Arr[Double] = ???
  def apply(x: Unit,xs: Unit*): Arr[Unit] = ???
}

object Test1 {
  implicit def arrToTrav[T](a: Arr[T]): Traversable[T] = ???

  def foo(t: Traversable[Any]) {}

  Arr("")        // okay
  foo(Arr(""))   // fails 2.10.3, ticket/8129


  Array("")      // okay
  foo(Array("")) // okay 2.10.3, fails ticket/8129
}

% qbin/scalac test/files/pos/t8219b.scala
test/files/pos/t8219b.scala:21: error: overloaded method value apply with alternatives:
  (x: Unit,xs: Unit*)Arr[Unit] <and>
  (x: Double,xs: Double*)Arr[Double] <and>
  (x: Float,xs: Float*)Arr[Float] <and>
  (x: Long,xs: Long*)Arr[Long] <and>
  (x: Int,xs: Int*)Arr[Int] <and>
  (x: Char,xs: Char*)Arr[Char] <and>
  (x: Short,xs: Short*)Arr[Short] <and>
  (x: Byte,xs: Byte*)Arr[Byte] <and>
  (x: Boolean,xs: Boolean*)Arr[Boolean]
 cannot be applied to (String)
  foo(Arr(""))   // fails 2.10.3, ticket/8129
      ^
test/files/pos/t8219b.scala:25: error: overloaded method value apply with alternatives:
  (x: Unit,xs: Unit*)Array[Unit] <and>
  (x: Double,xs: Double*)Array[Double] <and>
  (x: Float,xs: Float*)Array[Float] <and>
  (x: Long,xs: Long*)Array[Long] <and>
  (x: Int,xs: Int*)Array[Int] <and>
  (x: Char,xs: Char*)Array[Char] <and>
  (x: Short,xs: Short*)Array[Short] <and>
  (x: Byte,xs: Byte*)Array[Byte] <and>
  (x: Boolean,xs: Boolean*)Array[Boolean]
 cannot be applied to (String)
  foo(Array("")) // okay 2.10.3, fails ticket/8129
      ^
two errors found

% scalac-hash v2.10.3 test/files/pos/t8219b.scala
test/files/pos/t8219b.scala:21: error: overloaded method value apply with alternatives:
  (x: Unit,xs: Unit*)Arr[Unit] <and>
  (x: Double,xs: Double*)Arr[Double] <and>
  (x: Float,xs: Float*)Arr[Float] <and>
  (x: Long,xs: Long*)Arr[Long] <and>
  (x: Int,xs: Int*)Arr[Int] <and>
  (x: Char,xs: Char*)Arr[Char] <and>
  (x: Short,xs: Short*)Arr[Short] <and>
  (x: Byte,xs: Byte*)Arr[Byte] <and>
  (x: Boolean,xs: Boolean*)Arr[Boolean]
 cannot be applied to (String)
  foo(Arr(""))   // fails 2.10.3, ticket/8129
      ^
one error found
```

lrytz commented:
the `2.10.3` version looks dubious.

retronym commented:
During `isApplicable` for the generic apply method, we hit an ambiguous error when determining of the result type, `Array[?T]` is coercible to `Traversable[Any]`. The difference between `Arr` and `Arr` are the implicits in scope `Predef.{wrapLongArray, .. }`. More than one of these are deemed potentially applicable. 

Anyway, we rely on that ambiguity bubbling up into the outer context used in `Infer#tryTwice`, which resolves things on the second pass with by disabling views. Or maybe its the caller to `tryTwice` which attempts to recover from failure with wildcard expected type.

retronym commented:
Here's a slightly smaller minimization. 

``` scala
trait Arr[T]
object Arr {
  def apply[T](xs: T*)        : Arr[T]    = ???
  def apply(x: Long,xs: Long*): Arr[Long] = ???
}

object I {
  implicit def arrToTrav[T] (a: Arr[T])   : Traversable[T]    = ???  
  implicit def longArrToTrav(a: Arr[Long]): Traversable[Long] = ???
}

object Test {
  def foo(t: Traversable[Any]) {}

  object Okay {
    Arr("")

    import I.{ arrToTrav, longArrToTrav }
    foo(Arr(""))
  }

  object Fail {
    import I.arrToTrav
    foo(Arr("")) // 2.10.3: found String, expected Long
  }
}

```

retronym commented:
Lodged as [SI-8230](https://issues.scala-lang.org/browse/SI-8230).

I need to think a bit more about what to do for this bug. It's tempting to peel off the last commit, but I don't think its safe until we get to the bottom of SI-8230. Closing for now.
